GVFS/GVFS.Common/LocalUpgraderServices.cs

```diff
@@ -0,0 +1,152 @@
+
+namespace GVFS.Common
+{
+    public class LocalUpgraderServices
+    {
+        protected PhysicalFileSystem fileSystem;
+        protected ITracer tracer;
+
+        private const string ToolsDirectory = "Tools";
+        private static readonly string UpgraderToolName = GVFSPlatform.Instance.Constants.GVFSUpgraderExecutableName;
+        private static readonly string UpgraderToolConfigFile = UpgraderToolName + ".config";
+        private static readonly string[] UpgraderToolAndLibs =
+            {
+                UpgraderToolName,
+                UpgraderToolConfigFile,
+                "GVFS.Common.dll",
+                "GVFS.Platform.Windows.dll",
+                "Microsoft.Diagnostics.Tracing.EventSource.dll",
+                "netstandard.dll",
+                "System.Net.Http.dll",
+                "Newtonsoft.Json.dll"
+            };
+
+        public LocalUpgraderServices(ITracer tracer)
+        {
+            this.fileSystem = new PhysicalFileSystem();
+            this.tracer = tracer;
+        }
+
+        public string TempPath => LocalUpgraderServices.GetTempPath();
+
+        public static string GetTempPath()
+        {
+            return Path.Combine(
+                ProductUpgraderInfo.GetUpgradesDirectoryPath(),
+                "InstallerTemp");
+        }
+
+        public static bool TryCreateDirectory(string path, out Exception exception)
+        {
+            try
+            {
+                Directory.CreateDirectory(path);
+            }
+            catch (IOException e)
+            {
+                exception = e;
+                return false;
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                exception = e;
+                return false;
+            }
+
+            exception = null;
+            return true;
+        }
+
+        // TrySetupToolsDirectory -
+        // Copies GVFS Upgrader tool and its dependencies to a temporary location in ProgramData.
+        // Reason why this is needed - When GVFS.Upgrader.exe is run from C:\ProgramFiles\GVFS folder
+        // upgrade installer that is downloaded and run will fail. This is because it cannot overwrite
+        // C:\ProgramFiles\GVFS\GVFS.Upgrader.exe that is running. Moving GVFS.Upgrader.exe along with
+        // its dependencies to a temporary location inside ProgramData and running GVFS.Upgrader.exe
+        // from this temporary location helps avoid this problem.
+        public virtual bool TrySetupToolsDirectory(out string upgraderToolPath, out string error)
+        {
+            string rootDirectoryPath = ProductUpgraderInfo.GetUpgradesDirectoryPath();
+            string toolsDirectoryPath = Path.Combine(rootDirectoryPath, ToolsDirectory);
+            Exception exception;
+            if (TryCreateDirectory(toolsDirectoryPath, out exception))
+            {
+                string currentPath = ProcessHelper.GetCurrentProcessLocation();
+                error = null;

```

Comment:
See [this discussion](https://github.com/Microsoft/VFSForGit/pull/675/files/7ecafcfa101197c323607e748611fa610fa014fc#r246064887) about setting out-params by default at the beginning (tl;dr don't. let the compiler help you).